src/icp_v2.cc

```diff
@@ -670,38 +649,37 @@ icpHandleUdp(int sock, void *)
             if (xerrno != ECONNREFUSED && xerrno != EHOSTUNREACH)
 #endif
                 debugs(50, DBG_IMPORTANT, "icpHandleUdp: FD " << sock << " recvfrom: " << xstrerr(xerrno));
-
             break;
         }
 
         ++(*N);
         icpCount(buf, RECV, (size_t) len, 0);
         buf[len] = '\0';
-        debugs(12, 4, "icpHandleUdp: FD " << sock << ": received " <<
-               (unsigned long int)len << " bytes from " << from);
-
-#ifdef ICP_PACKET_DUMP
 
-        icpPktDump(buf);
-#endif
-
-        if ((size_t) len < sizeof(icp_common_t)) {
-            debugs(12, 4, "icpHandleUdp: Ignoring too-small UDP packet");
-            break;
-        }
+        icp_common_t pkt(buf, len);
 
-        icp_version = (int) buf[1]; /* cheat! */
+        debugs(12, 2, "ICP Client remote=" << from << " FD " << sock);
+        debugs(12, 2, "ICP Client REQUEST:\n---------\n" <<
+               icp_opcode_str[pkt.opcode] << " #" << pkt.reqnum << " ICP/" << pkt.version << std::endl <<
+               ":length: " << pkt.length << std::endl <<
+               ":flags: " << asHex(pkt.flags) <<
+               Debug::Extra << Raw("payload", (buf+sizeof(icp_common_t)), pkt.length).minLevel(DBG_DATA).hex() <<
```

Comment:
Given that this is a multiline output that follows very special formatting, the Debug::Extra trick does not work well here -- the payload dump gets misaligned. There is also a problem with Raw() printing an extra space when labeled. I suggest doing something like this to fix those problems:

```C++
":flags: " << asHex(pkt.flags) << '\n' <<
":payload:" << Raw("raw", (buf+sizeof(icp_common_t)), pkt.length).minLevel(DBG_DATA).hex() << '\n' <<
"----------");
```

We can also use "bytes" (or something else) instead of "raw" as the Raw() label. We could use nullptr label instead (and add a space after `:` before Raw()), but that will hide the payload size. Your call. And please test this. I may be missing some formatting nuances.